### PR TITLE
Add usersimcrs package

### PR DESCRIPTION
**Pull request by NoB0**
This PR add the package [usersimcrs](https://github.com/iai-group/UserSimCRS). Below is the modified and added files compared to original package.

Modified:
- `interaction_model.py` specifically agenda update

Added:
- `simulation_domain.py`
- `moviebot_rl_sim.py`

### CODE REVIEW

File: usersimcrs/domain/simulation_domain.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+"""Domain knowledge for the simulation."""
```
Comment (kbalog) [inline usersimcrs/domain/simulation_domain.py:1]: Would you mind expanding the docstring and adding that the domain specifies the slots that are used for elicitation and inquiry and add a toy example?

File: usersimcrs/domain/simulation_domain.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    def get_slot_names(self) -> List[str]:
+        """Returns the list of slot names.
+
+        Returns:
+            List of slot names.
+        """
+        return list(self._config.get("slot_names").keys())
+
+    def get_slot_names_elicitation(self) -> List[str]:
+        """Returns the list of slot names.
```
Comment (kbalog) [inline usersimcrs/domain/simulation_domain.py:28]: Nit: slot names for elicitation

File: usersimcrs/simulator/agenda_based/interaction_model.py [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+            .get("expected_agent_intents")
+        ) or []
+
+        logger.debug(f"Agent intent: {agent_intent}\n")
+
+        if not self._agenda:
+            self._current_intent = self.INTENT_STOP  # type: ignore[attr-defined] # noqa
+            return
+
+        # If agent replies in an expected intent, then pop the next intent from
```
Comment (kbalog) [inline usersimcrs/simulator/agenda_based/interaction_model.py:346]: Nit: replies _with_ an expected intent

File: usersimcrs/simulator/moviebot/moviebot_rl_sim.py [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+        self._domain = domain
+        self._item_collection = item_collection
+        self._ratings = ratings
+        self.initialize_goal()
+
+    def _get_information_need(self) -> Dict[str, str]:
+        """Gets the information need for the goal.
+
+        Returns:
+            Information need as a dictionary of slot value pairs.
```
Comment (kbalog) [inline usersimcrs/simulator/moviebot/moviebot_rl_sim.py:66]: Nit: slot-value

File: usersimcrs/simulator/moviebot/moviebot_rl_sim.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+        self._nlu = nlu
+        self._nlg = nlg
+        self._domain = domain
+        self._item_collection = item_collection
+        self._ratings = ratings
+        self.initialize_goal()
+
+    def _get_information_need(self) -> Dict[str, str]:
+        """Gets the information need for the goal.
+
```
Comment (kbalog) [inline usersimcrs/simulator/moviebot/moviebot_rl_sim.py:64]: Can you explain the purpose of this method (similarly how it's done for `initialize_goal`)?

File: usersimcrs/simulator/moviebot/moviebot_rl_sim.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+        item.
+        For example: the user wants a comedy movie and always accept the
+        recommendation.
+
+        Returns:
+            User goal.
+        """
+        self._goal = dict()
+        self._goal["acceptance_probability"] = random.random()
+        # Define slot value pairs for the goal.
```
Comment (kbalog) [inline usersimcrs/simulator/moviebot/moviebot_rl_sim.py:92]: Nit: slot-value (everywhere)

File: usersimcrs/simulator/moviebot/moviebot_rl_sim.py [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+        Return:
+            Response intent and annotations.
+        """
+        elicited_slot = (
+            slot
+            if slot
+            else random.choice(self._domain.get_slot_names_elicitation())
+        )
+        # During training of the slot annotator, a slot's name and value can be
+        # the almost the same, e.g., (GENRE, genres). In that case, value does
```
Comment (kbalog) [inline usersimcrs/simulator/moviebot/moviebot_rl_sim.py:157]: Nit: "the almost..." => drop the

File: usersimcrs/simulator/moviebot/moviebot_rl_sim.py [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+        """
+        response_slot = None
+        response_slot = random.choice(self._domain.get_slot_names_inquiry())
+        return [Annotation(slot=response_slot, value="")]
+
+    def _retrieve_last_slot_value_pair(self) -> Annotation:
+        """Retrieves the last slot value pair from simulator's previous
+        utterances.
+
+        If not slot value pair is found, select a random one.
```
Comment (kbalog) [inline usersimcrs/simulator/moviebot/moviebot_rl_sim.py:283]: Nit: If no slot-value...

File: usersimcrs/simulator/moviebot/moviebot_rl_sim.py [hunk 9]
```diff
@@ -0,0 +1,422 @@
+"""User simulator to interact with MovieBot agent during reinforcement
```
Comment (kbalog) [inline usersimcrs/simulator/moviebot/moviebot_rl_sim.py:1]: Can you expand it with a high-level description of how the simulator works (how goal/info need is generated and how agenda updates work)?

**Review by kbalog — APPROVED** [review]
LGTM. (Comments only concern docstrings, you may merge the PR once those are addressed.)

File: usersimcrs/domain/simulation_domain.py (hunk 2 above)
Comment (NoB0) [reply usersimcrs/domain/simulation_domain.py:28]: Done.

File: usersimcrs/domain/simulation_domain.py (hunk 1 above)
Comment (NoB0) [reply usersimcrs/domain/simulation_domain.py:1]: Done.

File: usersimcrs/simulator/moviebot/moviebot_rl_sim.py (hunk 4 above)
Comment (NoB0) [reply usersimcrs/simulator/moviebot/moviebot_rl_sim.py:66]: Done.

File: usersimcrs/simulator/moviebot/moviebot_rl_sim.py (hunk 6 above)
Comment (NoB0) [reply usersimcrs/simulator/moviebot/moviebot_rl_sim.py:92]: Done.

File: usersimcrs/simulator/moviebot/moviebot_rl_sim.py (hunk 7 above)
Comment (NoB0) [reply usersimcrs/simulator/moviebot/moviebot_rl_sim.py:157]: Done.

File: usersimcrs/simulator/moviebot/moviebot_rl_sim.py (hunk 8 above)
Comment (NoB0) [reply usersimcrs/simulator/moviebot/moviebot_rl_sim.py:283]: Done.

File: usersimcrs/simulator/agenda_based/interaction_model.py (hunk 3 above)
Comment (NoB0) [reply usersimcrs/simulator/agenda_based/interaction_model.py:346]: Done.

File: usersimcrs/simulator/moviebot/moviebot_rl_sim.py (hunk 5 above)
Comment (NoB0) [reply usersimcrs/simulator/moviebot/moviebot_rl_sim.py:64]: Done.

File: usersimcrs/simulator/moviebot/moviebot_rl_sim.py (hunk 9 above)
Comment (NoB0) [reply usersimcrs/simulator/moviebot/moviebot_rl_sim.py:1]: Done.

**Pull request merged**